File: dart/day5_part1_2018.dart

import 'dart:io';

void main() {
  String polymer = File('input.txt').readAsStringSync().trim();
  
  String reactedPolymer = reactPolymer(polymer);
  
  print(reactedPolymer.length);
}

String reactPolymer(String polymer) {
  List<String> units = polymer.split('');
  List<String> result = [];
  
  for (String unit in units) {
    if (result.isNotEmpty && willReact(result.last, unit)) {
      result.removeLast();
    } else {
      result.add(unit);
    }
  }
  
  return result.join('');
}

bool willReact(String unit1, String unit2) {
  return unit1 != unit2 && unit1.toLowerCase() == unit2.toLowerCase();
}

File: typescript/day5_part1_2018.ts
import * as fs from 'fs';

const readInput = (filePath: string): string => {
    return fs.readFileSync(filePath, 'utf-8').trim();
};

const reactPolymer = (polymer: string): number => {
    const stack: string[] = [];
    
    for (const unit of polymer) {
        const lastUnit = stack[stack.length - 1];
        if (lastUnit && lastUnit !== unit && lastUnit.toLowerCase() === unit.toLowerCase()) {
            stack.pop();
        } else {
            stack.push(unit);
        }
    }
    
    return stack.length;
};

const main = () => {
    const polymer = readInput('input.txt');
    const result = reactPolymer(polymer);
    console.log(result);
};

main();